<a href="https://colab.research.google.com/github/thiagocunha88/DataScience-Challenge_DS_Alura/blob/master/Challenge_Alura_DS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Challenge Alura - Reduzir a taxa de evasão dos clientes (Churn Rate)
## Analisar os clientes para identificar um padrão que determine aqueles que possuem maior chance de deixar de consumir o serviço, ou deixar de ser cliente da empresa

## Para iniciar os estudos, inicia-se a importação das bibliotecas básicas para tal e o dataset a ser analisado, entendendo que o primeiro passo é entender os dados a serem utilizados.

### Importando bibliotecas

Para iniciar uma análise exploratória é necessário transformar e visualizar os dados.

Foram selecionadas duas bibliotecas muito comuns e de bom desempenho para isto, o pandas e numpy.

Posteriormente serão importadas outras bibliotecas, principalmente quando forem geradas visualizações gráficas, mas isso pode aguardar.

In [17]:
import pandas as pd
import numpy as np

pd.options.plotting.backend = 'plotly'

### Importando o dataset e tendo a primeira visualização

Como a base dados a ser estudada está em formato json, optei por utilizar o pd.read_json() para fazer a leitura, principalmente porque apresenta os dados em formato tabular, facilitando as primeiras impressões e possibilitando nortear os próximos passos.

In [18]:
df = pd.read_json('https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json')
display(df.head(2))

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...

Os dados importados geraram uma tabela, porém, a mesma apresenta colunas ainda em formato json, dificultando o entendimento e impossibilitando análises estatísticas básicas de maneira simples.

Sendo assim, será necessário iniciar a transformação dos dados.

### Transformando os dados de json para dataframe pandas.

Todas as colunas a partir da Churn estavam em formato json ainda, dificultando a visualização completa bem como cálculos e análises a serem realizados.

Foi efetuada a conversão dos dados, primeiramente, criando dataframes separados para cada coluna a ser tratada (customer, phone, internet e account), utilizando a função pd.json_normalize().

Na coluna account foi necessário uma atenção maior, já que havia uma cadeia de json (Charges), situação que foi tratada pela função pd.json_normalize().

In [19]:
# Separando valores que estão em formato json, e retornando para o dataframe
customer = pd.json_normalize(df.customer)
phone = pd.json_normalize(df.phone)
internet = pd.json_normalize(df.internet)
account = pd.json_normalize(df.account)
df_base = pd.concat([df.customerID, df.Churn, customer, phone, internet, account], axis = 1)

### Fazendo a tradução dos cabeçalhos

A base de dados está toda em inglês, e para facilitar o entendimento do que cada coluna significa, foi efetuada a tradução das mesmas.

Posteriormente será analisada a necessidade de traduzir também os outros dados do dataset.

In [20]:
# Traduzindo nomes das colunas manualmente
df_base.columns = ['clienteID', 
                   'Churn', 
                   'genero_cliente', 
                   'mais_que_65_anos', 
                   'possui_parceiro_a', 
                  'possui_dependentes', 
                  'meses_contrato',
                  'servico_telefonico',
                  'multiplas_linhas',
                  'servico_internet',
                  'assinatura_seguranca',
                  'assinatura_backup',
                  'assinatura_protecao_dispositivo',
                  'assinatura_suporte',
                  'assinatura_TV',
                  'assinatura_streaming',
                  'contrato',
                  'fatura_digital',
                  'forma_pagamento',
                  'total_servicos_mes',
                  'total_gasto']

df_base.head()

clienteID Churn genero_cliente  mais_que_65_anos possui_parceiro_a  \
0  0002-ORFBO    No         Female                 0               Yes   
1  0003-MKNFE    No           Male                 0                No   
2  0004-TLHLJ   Yes           Male                 0                No   
3  0011-IGKFF   Yes           Male                 1               Yes   
4  0013-EXCHZ   Yes         Female                 1               Yes   

  possui_dependentes  meses_contrato servico_telefonico multiplas_linhas  \
0                Yes               9                Yes               No   
1                 No               9                Yes              Yes   
2                 No               4                Yes               No   
3                 No              13                Yes               No   
4                 No               3                Yes               No   

  servico_internet  ... assinatura_backup assinatura_protecao_dispositivo  \
0              DSL  ...               Yes                              No   
1              DSL  ...                No                              No   
2      Fiber optic  ...                No                             Yes   
3      Fiber optic  ...               Yes                             Yes   
4      Fiber optic  ...                No                              No   

  assinatura_suporte assinatura_TV assinatura_streaming        contrato  \
0                Yes           Yes                   No        One year   
1                 No            No                  Yes  Month-to-month   
2                 No            No                   No  Month-to-month   
3                 No           Yes                  Yes  Month-to-month   
4                Yes           Yes                   No  Month-to-month   

  fatura_digital   forma_pagamento total_servicos_mes  total_gasto  
0            Yes      Mailed check               65.6        593.3  
1             No      Mailed check               59.9        542.4  
2            Yes  Electronic check               73.9       280.85  
3            Yes  Electronic check               98.0      1237.85  
4            Yes      Mailed check               83.9        267.4  

[5 rows x 21 columns]

Agora que foi possível entender do que se trata cada coluna é importante verificar se os dados existentes nelas são consistentes e possibilitam efetuar análises.

Primeiramente é necessário verificar se há dados faltantes, e se eles impossibilitarão análises posterior.

O pandas tem a função .info() que retorna os tipos de cada coluna, além de informar se há dados nulos nelas.

In [21]:
df_base.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   clienteID                        7267 non-null   object 
 1   Churn                            7267 non-null   object 
 2   genero_cliente                   7267 non-null   object 
 3   mais_que_65_anos                 7267 non-null   int64  
 4   possui_parceiro_a                7267 non-null   object 
 5   possui_dependentes               7267 non-null   object 
 6   meses_contrato                   7267 non-null   int64  
 7   servico_telefonico               7267 non-null   object 
 8   multiplas_linhas                 7267 non-null   object 
 9   servico_internet                 7267 non-null   object 
 10  assinatura_seguranca             7267 non-null   object 
 11  assinatura_backup                7267 non-null   object 
 12  assinatura_protecao_

A coluna total_gasto apresenta o tipo 'object', porém, se trata de um valor numérico flutuante, ou float, e por isto se faz necessária a conversão do mesmo.

Utilizei a função pd.to_numeric para isto.

In [22]:
# Convertendo coluna total_gasto para float
df_base.total_gasto = pd.to_numeric(df_base.total_gasto, errors='coerce')
df_base.head(2)

clienteID Churn genero_cliente  mais_que_65_anos possui_parceiro_a  \
0  0002-ORFBO    No         Female                 0               Yes   
1  0003-MKNFE    No           Male                 0                No   

  possui_dependentes  meses_contrato servico_telefonico multiplas_linhas  \
0                Yes               9                Yes               No   
1                 No               9                Yes              Yes   

  servico_internet  ... assinatura_backup assinatura_protecao_dispositivo  \
0              DSL  ...               Yes                              No   
1              DSL  ...                No                              No   

  assinatura_suporte assinatura_TV assinatura_streaming        contrato  \
0                Yes           Yes                   No        One year   
1                 No            No                  Yes  Month-to-month   

  fatura_digital forma_pagamento total_servicos_mes  total_gasto  
0            Yes    Mailed check               65.6        593.3  
1             No    Mailed check               59.9        542.4  

[2 rows x 21 columns]

Após converter os dados, importante verificar os dados obtidos, pois na função to_numeric foi utilizado o argumento *errors = 'coerce'*. Este argumento faz com que os dados que tenham erros sejam covertidos em NaN, e estes podem atrapalhar a análise e montagem de gráficos.

In [23]:
df_base[df_base.total_gasto.isna()]

clienteID Churn genero_cliente  mais_que_65_anos possui_parceiro_a  \
975   1371-DWPAZ    No         Female                 0               Yes   
1775  2520-SGTTA    No         Female                 0               Yes   
1955  2775-SEFEE    No           Male                 0                No   
2075  2923-ARZLG    No           Male                 0               Yes   
2232  3115-CZMZD    No           Male                 0                No   
2308  3213-VVOLG    No           Male                 0               Yes   
2930  4075-WKNIU    No         Female                 0               Yes   
3134  4367-NUYAO    No           Male                 0               Yes   
3203  4472-LVYGI    No         Female                 0               Yes   
4169  5709-LVOEQ    No         Female                 0               Yes   
5599  7644-OMVMY    No           Male                 0               Yes   

     possui_dependentes  meses_contrato servico_telefonico  multiplas_linhas  \
975                 Yes               0                 No  No phone service   
1775                Yes               0                Yes                No   
1955                Yes               0                Yes               Yes   
2075                Yes               0                Yes                No   
2232                Yes               0                Yes                No   
2308                Yes               0                Yes               Yes   
2930                Yes               0                Yes               Yes   
3134                Yes               0                Yes               Yes   
3203                Yes               0                 No  No phone service   
4169                Yes               0                Yes                No   
5599                Yes               0                Yes                No   

     servico_internet  ...    assinatura_backup  \
975               DSL  ...                  Yes   
1775               No  ...  No internet service   
1955              DSL  ...                  Yes   
2075               No  ...  No internet service   
2232               No  ...  No internet service   
2308               No  ...  No internet service   
2930              DSL  ...                  Yes   
3134               No  ...  No internet service   
3203              DSL  ...                   No   
4169              DSL  ...                  Yes   
5599               No  ...  No internet service   

     assinatura_protecao_dispositivo   assinatura_suporte  \
975                              Yes                  Yes   
1775             No internet service  No internet service   
1955                              No                  Yes   
2075             No internet service  No internet service   
2232             No internet service  No internet service   
2308             No internet service  No internet service   
2930                             Yes                  Yes   
3134             No internet service  No internet service   
3203                             Yes                  Yes   
4169                             Yes                   No   
5599             No internet service  No internet service   

            assinatura_TV assinatura_streaming  contrato fatura_digital  \
975                   Yes                   No  Two year             No   
1775  No internet service  No internet service  Two year             No   
1955                   No                   No  Two year            Yes   
2075  No internet service  No internet service  One year            Yes   
2232  No internet service  No internet service  Two year             No   
2308  No internet service  No internet service  Two year             No   
2930                  Yes                   No  Two year             No   
3134  No internet service  No internet service  Two year             No   
3203                  Yes                   No  Two year            Yes   
4169                  Yes

Foram identificadas 11 linhas em que o total_gasto contém NaN, coincidindo com as linhas em que os clientes ainda não possuem um mês completo de utilização dos serviços (coluna meses_contrato = 0).

Será considerado o valor da coluna total_servicos_mes nestes casos, pois indica o valor gasto pelo cliente em todo o período de assinatura do mesmo (no caso, um mês incompleto).

In [24]:
df_base.loc[df_base.total_gasto.isna(), 'total_gasto'] = df_base.loc[df_base.total_gasto.isna(), 'total_servicos_mes']
df_base[df_base.meses_contrato == 0][['total_servicos_mes', 'total_gasto']]

total_servicos_mes  total_gasto
975                56.05        56.05
1775               20.00        20.00
1955               61.90        61.90
2075               19.70        19.70
2232               20.25        20.25
2308               25.35        25.35
2930               73.35        73.35
3134               25.75        25.75
3203               52.55        52.55
4169               80.85        80.85
5599               19.85        19.85

Achei interessante a ideia de criar uma função para validar os valores apresentados nas colunas para evitar a repetição de código.

Aproveitei pra criar uma lista de todos os valores únicos, para possibilitar uma tradução dos valores contidos na tabela, além dos valores das colunas (que já foram traduzidos).

In [25]:
df_base.loc[:,'Churn':].columns

Index(['Churn', 'genero_cliente', 'mais_que_65_anos', 'possui_parceiro_a',
       'possui_dependentes', 'meses_contrato', 'servico_telefonico',
       'multiplas_linhas', 'servico_internet', 'assinatura_seguranca',
       'assinatura_backup', 'assinatura_protecao_dispositivo',
       'assinatura_suporte', 'assinatura_TV', 'assinatura_streaming',
       'contrato', 'fatura_digital', 'forma_pagamento', 'total_servicos_mes',
       'total_gasto'],
      dtype='object')

In [26]:
def valores_unicos(df):
    lista_valores_unicos = []
    for coluna in df.loc[:,'Churn':].columns:
        if len(df[f'{coluna}'].unique()) > 5:
            print(f'{coluna} contém muitos valores únicos, analisar com maior detalhamento.')
        else:
            valores = df[f'{coluna}'].unique()
            lista_valores_unicos.extend(list(valores))
            print(f'{coluna}: ' + str(valores))
    lista_valores_unicos = list(set(lista_valores_unicos))
    return lista_valores_unicos

In [27]:
lista_valores_unicos = valores_unicos(df_base)

Churn: ['No' 'Yes' '']
genero_cliente: ['Female' 'Male']
mais_que_65_anos: [0 1]
possui_parceiro_a: ['Yes' 'No']
possui_dependentes: ['Yes' 'No']
meses_contrato contém muitos valores únicos, analisar com maior detalhamento.
servico_telefonico: ['Yes' 'No']
multiplas_linhas: ['No' 'Yes' 'No phone service']
servico_internet: ['DSL' 'Fiber optic' 'No']
assinatura_seguranca: ['No' 'Yes' 'No internet service']
assinatura_backup: ['Yes' 'No' 'No internet service']
assinatura_protecao_dispositivo: ['No' 'Yes' 'No internet service']
assinatura_suporte: ['Yes' 'No' 'No internet service']
assinatura_TV: ['Yes' 'No' 'No internet service']
assinatura_streaming: ['No' 'Yes' 'No internet service']
contrato: ['One year' 'Month-to-month' 'Two year']
fatura_digital: ['Yes' 'No']
forma_pagamento: ['Mailed check' 'Electronic check' 'Credit card (automatic)'
 'Bank transfer (automatic)']
total_servicos_mes contém muitos valores únicos, analisar com maior detalhamento.
total_gasto contém muitos valores úni

In [28]:
lista_valores_unicos = np.sort(lista_valores_unicos)
lista_valores_unicos

array(['', '0', '1', 'Bank transfer (automatic)',
       'Credit card (automatic)', 'DSL', 'Electronic check', 'Female',
       'Fiber optic', 'Mailed check', 'Male', 'Month-to-month', 'No',
       'No internet service', 'No phone service', 'One year', 'Two year',
       'Yes'], dtype='<U25')

Com a lista de valores únicos pronta, basta criar uma lista com os valores traduzidos e executar um .replace no dataframe inteiro.

Apenas para manter o padrão, criei novo dataframe inidicando que se trata de valores traduzidos.

In [29]:
lista_valores_unicos_traduzidos = [
                                   'Transferência bancária (automática)',
                                   'Cartão de crédito (automático)',
                                   'DSL',
                                   'Cheque eletrônico',
                                   'Feminino',
                                   'Fibra ótica',
                                   'Cheque por correio',
                                   'Masculino',
                                   'Mês a mês',
                                   'Não',
                                   'Sem serviço de internet',
                                   'Sem serviço telefônico',
                                   'Um ano',
                                   'Dois anos',
                                   'Sim'
                                   ]

df_base_traduzido = df_base.replace(lista_valores_unicos[3:], lista_valores_unicos_traduzidos)
df_base_traduzido.sample(5)

clienteID Churn genero_cliente  mais_que_65_anos possui_parceiro_a  \
2027  2862-JVEOY   Não      Masculino                 0               Não   
695   0988-AADSA   Não       Feminino                 0               Sim   
1843  2623-DRYAM   Não       Feminino                 0               Sim   
7227  9938-ZREHM   Não       Feminino                 0               Sim   
666   0946-FKYTX   Não      Masculino                 0               Não   

     possui_dependentes  meses_contrato servico_telefonico  \
2027                Não               6                Sim   
695                 Sim              72                Sim   
1843                Não              70                Sim   
7227                Não              37                Sim   
666                 Não              52                Não   

            multiplas_linhas servico_internet  ...        assinatura_backup  \
2027                     Não              Não  ...  Sem serviço de internet   
695                      Sim              DSL  ...                      Sim   
1843                     Sim      Fibra ótica  ...                      Não   
7227                     Sim              DSL  ...                      Sim   
666   Sem serviço telefônico              DSL  ...                      Sim   

     assinatura_protecao_dispositivo       assinatura_suporte  \
2027         Sem serviço de internet  Sem serviço de internet   
695                              Sim                      Sim   
1843                             Sim                      Não   
7227                             Não                      Sim   
666                              Não                      Não   

                assinatura_TV     assinatura_streaming   contrato  \
2027  Sem serviço de internet  Sem serviço de internet     Um ano   
695                       Não                      Sim  Dois anos   
1843                      Sim                      Sim     Um ano   
7227                      Não                      Sim     Um ano   
666                       Não                      Não     Um ano   

     fatura_digital                      forma_pagamento total_servicos_mes  \
2027            Não                   Cheque por correio              19.15   
695             Não  Transferência bancária (automática)              80.85   
1843            Sim                    Cheque eletrônico             101.75   
7227            Não                   Cheque por correio              72.10   
666             Não                   Cheque por correio              30.10   

      total_gasto  
2027       124.40  
695       5824.75  
1843      7069.30  
7227      2658.40  
666       1623.40  

[5 rows x 21 columns]

In [30]:
valores_unicos(df_base_traduzido)

Churn: ['Não' 'Sim' '']
genero_cliente: ['Feminino' 'Masculino']
mais_que_65_anos: [0 1]
possui_parceiro_a: ['Sim' 'Não']
possui_dependentes: ['Sim' 'Não']
meses_contrato contém muitos valores únicos, analisar com maior detalhamento.
servico_telefonico: ['Sim' 'Não']
multiplas_linhas: ['Não' 'Sim' 'Sem serviço telefônico']
servico_internet: ['DSL' 'Fibra ótica' 'Não']
assinatura_seguranca: ['Não' 'Sim' 'Sem serviço de internet']
assinatura_backup: ['Sim' 'Não' 'Sem serviço de internet']
assinatura_protecao_dispositivo: ['Não' 'Sim' 'Sem serviço de internet']
assinatura_suporte: ['Sim' 'Não' 'Sem serviço de internet']
assinatura_TV: ['Sim' 'Não' 'Sem serviço de internet']
assinatura_streaming: ['Não' 'Sim' 'Sem serviço de internet']
contrato: ['Um ano' 'Mês a mês' 'Dois anos']
fatura_digital: ['Sim' 'Não']
forma_pagamento: ['Cheque por correio' 'Cheque eletrônico' 'Cartão de crédito (automático)'
 'Transferência bancária (automática)']
total_servicos_mes contém muitos valores únicos, an

['',
 0,
 1,
 'Cheque eletrônico',
 'Cheque por correio',
 'Um ano',
 'DSL',
 'Não',
 'Cartão de crédito (automático)',
 'Sem serviço telefônico',
 'Feminino',
 'Sim',
 'Masculino',
 'Dois anos',
 'Mês a mês',
 'Transferência bancária (automática)',
 'Fibra ótica',
 'Sem serviço de internet']

In [31]:
display((df_base_traduzido.groupby('Churn').count()['clienteID']))
display((df_base_traduzido.groupby('Churn').count()['clienteID'] / df_base_traduzido.Churn.count() * 100))


Churn
        224
Não    5174
Sim    1869
Name: clienteID, dtype: int64

Churn
        3.082427
Não    71.198569
Sim    25.719004
Name: clienteID, dtype: float64

Com a informação dos valores únicos, foi possível perceber que a coluna Churn possui, além de Yes e No, valores vazios. Para ser mais preciso, 224 ocorrências, ou 3,08%.

O foco do estudo é endenter o comportamento de clientes usando como variável dependente o Churn, além do que, a quantidade de vazios não é expressiva e por isto não vejo sentido em manter as linhas em que este valore está vazio.

Para fazer isto, utilizei a função .drop.

In [32]:
df_base_traduzido.drop(df_base[df_base.Churn == ''].index, inplace = True)
df_base_traduzido.reset_index(drop = True)

clienteID Churn genero_cliente  mais_que_65_anos possui_parceiro_a  \
0     0002-ORFBO   Não       Feminino                 0               Sim   
1     0003-MKNFE   Não      Masculino                 0               Não   
2     0004-TLHLJ   Sim      Masculino                 0               Não   
3     0011-IGKFF   Sim      Masculino                 1               Sim   
4     0013-EXCHZ   Sim       Feminino                 1               Sim   
...          ...   ...            ...               ...               ...   
7038  9987-LUTYD   Não       Feminino                 0               Não   
7039  9992-RRAMN   Sim      Masculino                 0               Sim   
7040  9992-UJOEL   Não      Masculino                 0               Não   
7041  9993-LHIEB   Não      Masculino                 0               Sim   
7042  9995-HOTOH   Não      Masculino                 0               Sim   

     possui_dependentes  meses_contrato servico_telefonico  \
0                   Sim               9                Sim   
1                   Não               9                Sim   
2                   Não               4                Sim   
3                   Não              13                Sim   
4                   Não               3                Sim   
...                 ...             ...                ...   
7038                Não              13                Sim   
7039                Não              22                Sim   
7040                Não               2                Sim   
7041                Sim              67                Sim   
7042                Sim              63                Não   

            multiplas_linhas servico_internet  ... assinatura_backup  \
0                        Não              DSL  ...               Sim   
1                        Sim              DSL  ...               Não   
2                        Não      Fibra ótica  ...               Não   
3                        Não      Fibra ótica  ...               Sim   
4                        Não      Fibra ótica  ...               Não   
...                      ...              ...  ...               ...   
7038                     Não              DSL  ...               Não   
7039                     Sim      Fibra ótica  ...               Não   
7040                     Não              DSL  ...               Sim   
7041                     Não              DSL  ...               Não   
7042  Sem serviço telefônico              DSL  ...               Sim   

     assinatura_protecao_dispositivo assinatura_suporte assinatura_TV  \
0                                Não                Sim           Sim   
1                                Não                Não           Não   
2                                Sim                Não           Não   
3                                Sim                Não           Sim   
4                                Não                Sim           Sim   
...                              ...                ...           ...   
7038                             Não                Sim           Não   
7039                             Não                Não           Não   
7040                             Não                Não           Não   
7041                             Sim                Sim           Não   
7042                             Sim                Não           Sim   

     assinatura_streaming   contrato fatura_digital     forma_pagamento  \
0                     Não     Um ano            Sim  Cheque por correio   
1                     Sim  Mês a mês            Não  Cheque por correio   
2                     Não  Mês a mês            Sim   Cheque eletrônico   
3                     Sim  Mês a mês            Sim   Cheque eletrônico   
4                     Não  Mês a mês            Sim  Cheque por correio   
...                   ...        ...            ...                 ...   
7038                  Não     Um ano            Não  Cheque por correi

In [ ]:
# Criando coluna gastos_diarios, quando meses_contrato < 0, usar coluna total_servicos_mes,
# quando meses_contrato > 0 considera total_gasto
df_base['gastos_diarios'] = [df_base.total_gasto[i] / (df_base.meses_contrato[i] * 30) 
                             if df_base.meses_contrato[i] > 0 
                             else df_base.total_servicos_mes[i] / 30 
                             for i in df_base.index]

df_base.head(2)

### Validação de algumas colunas

In [ ]:
df_base.columns

In [ ]:
df_base.assinatura_protecao_dispositivo.unique()

In [ ]:
df_base[df_base.multiplas_linhas == 'No phone service'].shape[0]

In [ ]:
pd.crosstab(df_base.Churn, df_base.servico_internet)

In [ ]:
df_base.groupby('servico_internet').count()

In [ ]:
df_base.groupby('assinatura_seguranca').count()

In [ ]:
df_base.groupby('assinatura_backup').count()

In [ ]:
df_base.groupby('assinatura_protecao_dispositivo').count()

### Análise da variável target - Churn

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df_base.Churn.value_counts()

#### Avaliando por gênero

In [ ]:
df_genero = pd.crosstab(df_base.Churn, df_base.genero_cliente)
df_genero

In [ ]:
df_base.groupby('genero_cliente')['Churn'].count()

In [ ]:
df_genero = pd.crosstab(df_base.Churn, df_base.genero_cliente)

fig, ax = plt.subplots(figsize=(10, 8))

grafico = ax.bar(x = 'Churn', height=df_base.Churn, data = df_base)

for barra in grafico:
  # Pegando a altura(valor) e largura de cada barra
    altura = barra.get_height()
    largura = barra.get_width()

    '''
      Posição x(xpos) e y(ypos) de cada barra. Aqui temos três casos de y pos:
      1º - colocando o rótulo acima da barra
      2º - colocando o rótulo no meio da barra
      3º - colocando o rótulo na base da barra
      Para selecionar o desejado, é só descomentar o ypos que quer fazer uso.
    '''
    xpos = barra.get_x() + largura/2
    ypos = 1.01 * altura
    # ypos = barra.get_y() + altura/2
    # ypos = barra.get_y()

    # Adicionando o rótulo nas barras (x= coordenada x, y = coordenada y, s = texto desejado)
    ax.text(x = xpos, y = ypos, s = altura, ha='center', va='bottom', fontsize=14)

In [ ]:
fig, ax = plt.subplots(figsize=(10, 8))

# grafico = df_genero.plot(kind = 'bar')
grafico = ax.bar(x = df_genero.columns, height=df_genero.loc['Yes'], data = df_genero)

for barra in grafico:
  # Pegando a altura(valor) e largura de cada barra
    altura = barra.get_height()
    largura = barra.get_width()

    '''
      Posição x(xpos) e y(ypos) de cada barra. Aqui temos três casos de y pos:
      1º - colocando o rótulo acima da barra
      2º - colocando o rótulo no meio da barra
      3º - colocando o rótulo na base da barra
      Para selecionar o desejado, é só descomentar o ypos que quer fazer uso.
    '''
    xpos = barra.get_x() + largura/2
    ypos = 1.01 * altura
    # ypos = barra.get_y() + altura/2
    # ypos = barra.get_y()

    # Adicionando o rótulo nas barras (x= coordenada x, y = coordenada y, s = texto desejado)
    ax.text(x = xpos, y = ypos, s = altura, ha='center', va='bottom', fontsize=14)

#### Se acima de 65 anos

In [ ]:
df_analise = pd.crosstab(df_base.Churn, df_base.mais_que_65_anos.astype(str))

fig, ax = plt.subplots(figsize=(10, 8))

grafico = ax.bar(x = df_analise.columns, height=df_analise.loc['Yes'], data = df_analise.loc['Yes'])

for barra in grafico:
  # Pegando a altura(valor) e largura de cada barra
    altura = barra.get_height()
    largura = barra.get_width()

    '''
      Posição x(xpos) e y(ypos) de cada barra. Aqui temos três casos de y pos:
      1º - colocando o rótulo acima da barra
      2º - colocando o rótulo no meio da barra
      3º - colocando o rótulo na base da barra
      Para selecionar o desejado, é só descomentar o ypos que quer fazer uso.
    '''
    xpos = barra.get_x() + largura/2
    ypos = 1.01 * altura
    # ypos = barra.get_y() + altura/2
    # ypos = barra.get_y()

    # Adicionando o rótulo nas barras (x= coordenada x, y = coordenada y, s = texto desejado)
    ax.text(x = xpos, y = ypos, s = altura, ha='center', va='bottom', fontsize=14)